tsdate/util.py

```diff
@@ -121,7 +124,104 @@ def preprocess_ts(tree_sequence, minimum_gap=1000000, trim_telomeres=True):
         assert tree_sequence.num_sites == tree_sequence_trimmed.num_sites
         return provenance.record_provenance(
                 tree_sequence_trimmed, "preprocess_ts", minimum_gap=minimum_gap,
-                trim_telomeres=trim_telomeres, delete_intervals=delete_intervals)
+                remove_telomeres=remove_telomeres, delete_intervals=delete_intervals)
     else:
-        logger.info("No gaps to trim")
+        logger.info("No gaps to remove")
         return tree_sequence
+
+
+def nodes_time(tree_sequence, unconstrained=True):
+    nodes_age = tree_sequence.tables.nodes.time[:]
+    if unconstrained:
+        metadata = tree_sequence.tables.nodes.metadata[:]
+        metadata_offset = tree_sequence.tables.nodes.metadata_offset[:]
+        for index, met in enumerate(tskit.unpack_bytes(metadata, metadata_offset)):
+            if index not in tree_sequence.samples():
+                try:
+                    nodes_age[index] = json.loads(met.decode())["mn"]
+                except (KeyError, json.decoder.JSONDecodeError):
+                    raise ValueError("Tree Sequence must be tsdated with the\
+                                      Inside-Outside Method. Use unconstrained=False if\
+                                      not.")
+    return nodes_age
+
+
+def sites_time(tree_sequence, unconstrained=True, mutation_age="child",
+               ignore_multiallelic=True):
+    """
+    Returns the estimated time of the node of oldest mutation at each site.
+
+    :param TreeSequence tree_sequence: The input :class`tskit.TreeSequence`.
+    :param bool unconstrained: Use node ages which are unconstrained by site topology.
+        Only applies when the inside-outside algorithm is used. Default: "True"
+    :param str mutation_age: If "child", return the age of the child node of the oldest
+        mutation associated with a site. If "arithmetic", return the age of the oldest
+        mutation associated with each site, calculated using the arithmetic mean of
+        the parent and child nodes of the mutation. If "geometric", use the geometric
+        mean. "parent" returns the age of the node of the parent.
+        Default: "child"
+    :param bool ignore_multiallelic: If True, return age of oldest mutation associated
+        with an allele at multiallelic sites. If False, return tskit.UNKNOWN_TIME at any
+        multiallelic sites. (defaults to True).
+        Default: True
+    :return: An array of length tree_sequence.num_sites giving the estimated site time
+        of each site.
+    :rtype numpy.array
+
+    """
+    if tree_sequence.num_sites < 1:
+        raise ValueError("Invalid tree sequence: no sites present")
+    if mutation_age not in ["arithmetic", "geometric", "child"]:
```

Comment:
How about "parent" - have we tested this in the unit tests?

Reply:
I can add an option for "parent". I've tested the other three in the unit tests

Reply:
It's only that you mention it in the docstring...